### PR TITLE
Migration of isotropic power law hardening plasticity

### PR DESCRIPTION
**Pull request by sapitts**
### Relevant design change information

Isotropic power law hardening plasticity is migrated from solid mechanics to tensor mechanics.  

As part of the migration process, the iteration loop in `RecomputeRadialReturn` was changed from a while loop to a do-while loop.  Within Solid Mechanics, the output string was changed from using the dense stringstream method to the newer and easier to read `Moose::stringify` method.  The order of function calls in solid mechanics `IsotropicPlasticity` was changed to move `computeHardeningValue` above the residual calculation so that the effective plastic strain increment will evolve--I beat @bwspenc to this particular change.
### Test Cases

The test case for isotropic power law hardening was moved into the combined folder, and both the solid mechanics and tensor mechanics input files compare against the same gold file.  A few more tensor mechanics recompute radial return plasticity tests were changed to accommodate the new parameter name `temperature` which replaces the parameter name `temp` for consistency in parameter names. 

@dschwen Would you please review the Tensor Mechanics changes, and @jasondhales would you please review the changes to Solid Mechanics? 

Refs #5481

### CODE REVIEW

File: modules/tensor_mechanics/include/materials/RecomputeRadialReturnIsotropicPowerLawHardening.h [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+{
+public:
+  RecomputeRadialReturnIsotropicPowerLawHardening(const InputParameters & parameters);
+
+protected:
+  virtual void computeStressInitialize(Real effectiveTrialStress);
+  virtual void computeYieldStress();
+  virtual Real computeHardeningDerivative(Real scalar);
+
+  /// Power law hardening coefficients
```
Comment (dschwen) [inline modules/tensor_mechanics/include/materials/RecomputeRadialReturnIsotropicPowerLawHardening.h]: Just a remark, if you want the Doxygen comment to be applied to both lines you need to write

```
///@{ Power law hardening coefficients
Real _K;
Real _strain_hardening_exponent;
///@}
```

File: modules/tensor_mechanics/src/base/TensorMechanicsApp.C [hunk 2]
```diff
@@ -236,6 +237,7 @@ TensorMechanicsApp::registerObjects(Factory & factory)
   registerMaterial(ComputeReturnMappingStress);
   registerMaterial(RecomputeRadialReturn);
   registerMaterial(RecomputeRadialReturnIsotropicPlasticity);
+  registerMaterial(RecomputeRadialReturnIsotropicPowerLawHardening);
```
Comment (dschwen) [inline modules/tensor_mechanics/src/base/TensorMechanicsApp.C]: Sorry, this is not a new record in object name length. `MultiAppVariableValueSamplePostprocessorTransfer` still holds that one with 48 characters (yours only has 47... wanna try again?).

File: modules/tensor_mechanics/src/materials/RecomputeRadialReturnIsotropicPowerLawHardening.C [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+
+RecomputeRadialReturnIsotropicPowerLawHardening::RecomputeRadialReturnIsotropicPowerLawHardening(const InputParameters & parameters) :
+    RecomputeRadialReturnIsotropicPlasticity(parameters),
+    _K(parameters.get<Real>("strength_coefficient")),
+    _strain_hardening_exponent(parameters.get<Real>("strain_hardening_exponent"))
+{
+  if (_strain_hardening_exponent <= 0.0  || _strain_hardening_exponent >= 1.0)
+    mooseError("Verify the value of the strain hardening exponent, which should be a value between 0.0 and 1.0");
```
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/RecomputeRadialReturnIsotropicPowerLawHardening.C]: drop this check and use a range checked parameter instead!

File: modules/tensor_mechanics/src/materials/RecomputeRadialReturnIsotropicPowerLawHardening.C [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+  Real slope = std::pow(stress_delta, (1.0 / _strain_hardening_exponent - 1.0)) / _strain_hardening_exponent * 1.0 / std::pow(_K, 1.0 / _strain_hardening_exponent);
+  slope -= 1.0 / _youngs_modulus;
+
+  return 1.0 / slope;
+}
+
+void
+RecomputeRadialReturnIsotropicPowerLawHardening::computeYieldStress()
+{
+  // Pull in the Lame lambda, and caculate E
```
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/RecomputeRadialReturnIsotropicPowerLawHardening.C]: Should we try making this `Lamé`. Our minimum compiler requirements should be able to handle this.

File: modules/tensor_mechanics/src/materials/RecomputeRadialReturnIsotropicPowerLawHardening.C [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+  Real lambda = getIsotropicLameLambda();
+  _youngs_modulus = _shear_modulus * (3.0 * lambda + 2 * _shear_modulus) / ( lambda + _shear_modulus);
+
+  // Then solve for yield stress using equation from the header file
+  _yield_stress = std::pow(_K / pow(_youngs_modulus, _strain_hardening_exponent), 1.0 / (1.0 - _strain_hardening_exponent));
```
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/RecomputeRadialReturnIsotropicPowerLawHardening.C]: there is a `pow` without `std::`

File: modules/tensor_mechanics/src/materials/RecomputeRadialReturnIsotropicPowerLawHardening.C [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+  if (_yield_stress <= 0.0)
+    mooseError("The yield stress must be greater than zero, but during the simulation your yield stress became less than zero.");
+}
+
+Real
+RecomputeRadialReturnIsotropicPowerLawHardening::getIsotropicLameLambda()
+{
+  Real lame_lambda = _elasticity_tensor[_qp](0,0,1,1);
```
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/RecomputeRadialReturnIsotropicPowerLawHardening.C]: const

**Comment by dschwen** [comment]
:+1: looks good to me

**Comment by jasondhales** [comment]
👍 looks good to me too

**Pull request merged**